src/qibo/gates/abstract.py

```diff
@@ -359,18 +360,103 @@ def controlled_by(self, *qubits: int) -> "Gate":
             self.control_qubits = qubits
         return self
 
-    def decompose(self, *free) -> List["Gate"]:
+    def _base_decompose(self, *free, use_toffolis=True) -> List["Gate"]:
+        """Base decomposition for gates.
+
+        Returns a list containing the gate itself. Should be overridden by
+        subclasses that support decomposition to simpler gates.
+
+        Args:
+            free: Ids of free qubits to use for the gate decomposition.
+            use_toffolis: If ``True`` the decomposition contains only ``TOFFOLI`` gates.
+                If ``False`` a congruent representation is used for ``TOFFOLI`` gates.
+                See :class:`qibo.gates.TOFFOLI` for more details on this representation.
+
+        Returns:
+            list: Synthesis of the original gate in another gate set.
+        """
+        return [self.__class__(*self.init_args, **self.init_kwargs)]
+
+    @staticmethod
+    def _gates_cancel(g1, g2):
+        """Check if same gate type, same targets, same controls"""
+        if g1.__class__ != g2.__class__:
+            return False
+
+        if g1.target_qubits != g2.target_qubits:
+            return False
+        if (
+            getattr(g1, "is_controlled_by", True)
+            and g1.control_qubits != g2.control_qubits
+        ):
```

Comment:
```suggestion
        if (
            g1.control_qubits != g2.control_qubits
        ):
```
I think you only need to check the control qubits, as we said below a CNOT has ``is_controlled_by==False`` and thus would not enter here.